File: 2025/go/day-7/part-1.go
package day7

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const (
	FILE_PATH = "../inputs/day-7/part-1/input.txt"
)

func generateBeamList(line string) []bool {
	result := make([]bool, len(line))
	for i, r := range line {
		if r == 'S' {
			result[i] = true
			continue
		}
		result[i] = false
	}
	return result
}

func progressSplitters(line string, beamList []bool) (newBeamList []bool, splitCount int) {
	newBeamList = make([]bool, len(line))
	splitCount = 0
	for i, r := range line {
		if r == '^' && beamList[i] {
			if i-1 >= 0 {
				newBeamList[i-1] = true
			}
			if i+1 < len(line) {
				newBeamList[i+1] = true
			}
			newBeamList[i] = false
			splitCount++
			continue
		}
		newBeamList[i] = newBeamList[i] || beamList[i]
	}
	return newBeamList, splitCount
}

func formatBeamList(beamList []bool) string {
	result := ""
	for _, b := range beamList {
		if b {
			result += "|"
		} else {
			result += "."
		}
	}
	return result
}

func Day7Part1() {
	file, err := os.Open(FILE_PATH)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	shouldPrint := false

	splitCount := 0
	beamList := make([]bool, 0)
	idx := 0

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if idx == 0 {
			beamList = generateBeamList(line)
			if shouldPrint {
				fmt.Println(line)
				fmt.Println(formatBeamList(beamList))
			}
		} else if idx%2 == 0 {
			newBeamList, newSplitCount := progressSplitters(line, beamList)
			splitCount += newSplitCount
			beamList = newBeamList
			if shouldPrint {
				fmt.Println(line)
				fmt.Println(formatBeamList(beamList))
			}
		}
		idx++
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		os.Exit(1)
	}

	solution := splitCount

	fmt.Printf("Solution: %d\n", solution)

}

File: 2025/ts/src/day-7/part-1.ts
import * as fs from 'fs';
import * as readline from 'readline';
import { PATH } from './common';

function generateBeamList(line: string): boolean[] {
  const result: boolean[] = [];
  for (const char of line) {
    result.push(char === 'S');
  }
  return result;
}

function progressSplitters(line: string, beamList: boolean[]): [boolean[], number] {
  const newBeamList: boolean[] = new Array(line.length).fill(false);
  let splitCount = 0;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    if (char === '^' && beamList[i]) {
      if (i - 1 >= 0) {
        newBeamList[i - 1] = true;
      }
      if (i + 1 < line.length) {
        newBeamList[i + 1] = true;
      }
      newBeamList[i] = false;
      splitCount++;
      continue;
    }
    newBeamList[i] = newBeamList[i] || beamList[i];
  }

  return [newBeamList, splitCount];
}

function formatBeamList(beamList: boolean[]): string {
  let result = '';
  for (const b of beamList) {
    result += b ? '|' : '.';
  }
  return result;
}

export async function day7Part1(): Promise<void> {
  try {
    const fileStream = fs.createReadStream(PATH);
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    const shouldPrint = false;
    let splitCount = 0;
    let beamList: boolean[] = [];
    let idx = 0;

    for await (const line of rl) {
      const trimmedLine = line.trim();

      if (idx === 0) {
        beamList = generateBeamList(trimmedLine);
        if (shouldPrint) {
          console.log(trimmedLine);
          console.log(formatBeamList(beamList));
        }
      } else if (idx % 2 === 0) {
        const [newBeamList, newSplitCount] = progressSplitters(trimmedLine, beamList);
        splitCount += newSplitCount;
        beamList = newBeamList;
        if (shouldPrint) {
          console.log(trimmedLine);
          console.log(formatBeamList(beamList));
        }
      }

      idx++;
    }

    const solution = splitCount;
    console.log(`Solution: ${solution}`);
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}